tasks/service/impl/src/main/java/org/projectnessie/nessie/tasks/service/impl/TasksServiceImpl.java

```diff
@@ -0,0 +1,514 @@
+    // asynchronous task handling improves the implementation.
+    // TODO using `fetchObj()` would be wrong here, because it is *synchronous* and can block.
+    //  Options:
+    //  a) remove this "optimization"
+    //  b) add a `getObjIfCached(ObjId)` --> chosen for now
+    //  c) add a `fetchObjAsync()`, but adding async variants to all database implementations will
+    //     be tricky
+    Obj obj = persist.getIfCached(taskRequest.objId());
+    if (obj != null) {
+      TaskObj taskObj = (TaskObj) obj;
+      TaskStatus status = taskObj.taskState().status();
+      switch (status) {
+        case FAILURE:
+          metrics.taskHasFinalFailure();
+          return failedStage(taskRequest.behavior().stateAsException(taskObj));
+        case SUCCESS:
+          metrics.taskHasFinalSuccess();
+          return completedStage(taskObj);
+        default:
+          // task object exists but has a non-final state, handle it asynchronously
+          checkState(!status.isFinal(), "Expect non-final task status");
+          break;
+      }
+    }
+
+    // Ensure that only one "base" completable future exists for each obj-id.
+    TaskKey taskKey = TaskKey.taskKey(persist.config().repositoryId(), objId);
+
+    // The shutdown-check can be racy, if shutdown() is called after the `if` but before the access
+    // to `currentTasks`. But since `shutdown()` is usually only relevant for tests, that trade-off
+    // is acceptable.
+    if (shutdown) {
+      return CompletableFuture.failedStage(
+          new IllegalStateException("Tasks service already shutdown"));
+    }
+
+    return currentTasks.computeIfAbsent(
+        taskKey,
+        id -> {
+          metrics.startNewTaskController();
+          ExecParams execParams = new ExecParams(persist, taskRequest);
+          LOGGER.trace("{}: Starting new local task controller for {}", name, execParams);
+          async.call(() -> tryLocal(execParams));
+          return execParams.localResultFuture;
+        });
+  }
+
+  private void finalResult(ExecParams params, TaskObj result) {
+    try {
+      params.localResultFuture.complete(result);
+    } finally {
+      removeFromCurrentTasks(params);
+    }
+  }
+
+  private void finalFailure(ExecParams params, Throwable t) {
+    try {
+      params.localResultFuture.completeExceptionally(t);
+    } finally {
+      removeFromCurrentTasks(params);
+    }
+  }
+
+  private void removeFromCurrentTasks(ExecParams params) {
+    TaskKey taskKey = TaskKey.taskKey(params.persist.config().repositoryId(), params.objId());
+    currentTasks.remove(taskKey);
+  }
+
+  private void tryLocal(ExecParams params) {
+    try {
+      metrics.taskAttempt();
+      LOGGER.trace("{}: Task evaluation attempt for {}", name, params);
+
+      TaskObj obj = (TaskObj) params.persist.fetchObj(params.objId());
+      // keep in mind: `obj` might be a locally cached instance that is not in sync w/ the
```

Comment:
In this case shouldn't we "refresh" the task object and force a fetch from the DB?

Reply:
It might be stale. Cache-expiration however should make it pretty safe though.

Reply:
And it would be refreshed implicitly in the "next round"